test/gtest/uct/ib/test_ib_event.cc

```diff
@@ -15,47 +15,186 @@ extern "C" {
+
+    class qp {
+    public:
+        qp(entity &e) : m_e(e) {}
+        virtual ~qp() {}
+
+        virtual uint32_t qp_num() const = 0;
+        virtual void to_err()           = 0;
+
+        struct ibv_ah_attr ah_attr() const {
+            uct_ib_iface_t *iface = ucs_derived_of(m_e.iface(), uct_ib_iface_t);
+            struct ibv_ah_attr result = {};
+
+            uct_ib_iface_fill_ah_attr_from_gid_lid(iface,
+                    uct_ib_iface_port_attr(iface)->lid, &iface->gid_info.gid,
+                    iface->gid_info.gid_index, 0, &result);
+            return result;
+        }
+
+        enum ibv_mtu path_mtu() const {
+            uct_ib_iface_t *iface = ucs_derived_of(m_e.iface(), uct_ib_iface_t);
+
+            return iface->config.path_mtu;
+        }
+
+    protected:
+        entity &m_e;
+    };
+
+    struct event_ctx {
+        uct_ib_async_event_wait_t super;
+        volatile bool             got;
+    };
+
+    static void last_wqe_check_cb(uct_ib_async_event_wait_t *arg) {
+        event_ctx *event = ucs_derived_of(arg, event_ctx);
+        event->got       = true;
+    }
+
+    enum {
+        RC_MLX5,
+        RC_VERBS
+    };
+
+    virtual qp *create_qp(entity &e, int type) = 0;
+
+    void init_qp(entity &e) {
+        const resource *r = dynamic_cast<const resource*>(GetParam());
+        ucs_assert_always(r != NULL);
+
+        if (r->tl_name == "rc_mlx5") {
+            m_qp.reset(create_qp(e, RC_MLX5));
+        } else {
+            m_qp.reset(create_qp(e, RC_VERBS));
+        }
+    }
+
+    uct_ib_device_t *dev(entity &e) {
+        return &ucs_derived_of(e.md(), uct_ib_md_t)->dev;
+    }
+
+    bool wait_for_last_wqe_event(entity &e, bool before, int delay_usec = 0) {
+        uint32_t qp_num = m_qp->qp_num();
+        ucs_time_t deadline;
+        ucs_status_t status;
+        int ret;
+
+        m_event.got       = false;
+        m_event.super.cb  = last_wqe_check_cb;
+        m_event.super.cbq = &e.worker()->progress_q;
+
+        if (before) {
+            /* move QP to error state before scheduling event callback */
+            m_qp->to_err();
+            usleep(1000);
+        }
+
+        /* schedule event callback */
+        status = uct_ib_device_async_event_wait(dev(e),
+                IBV_EVENT_QP_LAST_WQE_REACHED, qp_num, &m_event.super);
+        ASSERT_UCS_OK_OR_INPROGRESS(status);
+        if (status == UCS_OK) {
+            /* event already arrived */
+            return true;
+        }
+
+        if (!before) {
+            /* move QP to error state after scheduling event callback */
+            m_qp->to_err();
+        }
+
+        if (delay_usec) {
+            /* wait for last_wqe event arrival */
+            usleep(delay_usec);
+            /* check if event was added to progress queue */
+            ucs_spin_lock(&dev(e)->async_event_lock);
+            ret = m_event.super.cb_id != UCS_CALLBACKQ_ID_NULL;
+            ucs_spin_unlock(&dev(e)->async_event_lock);
```

Comment:
why does it matter?

Reply:
to check that was large enough delay

Reply:
maybe call progress after calling wait_for_last_wqe_event?
IMO try to avoid access to internal structs